### DISCUSSION TITLE
求助：treeTable 异步加载 如何实现按照一定的顺序展开树

### DISCUSSION THREAD

**Discussion by chenh2h** [layui/layui#2232] [问答]
想要实现的功能：
点击按钮，按照读取到的顺序依次展开树
但实际执行时展开第三个节点时会报错
![image](https://github.com/user-attachments/assets/d64d2c35-b79f-42f4-a42e-72ef7a2e1a87)

    $('#load_last_path').click(function(){
        treeTable.expandNode('blade_testCase_tree', {index: '0',
            expandFlag: true,
            callbackFlag:false});
        treeTable.expandNode('blade_testCase_tree', {index: '0-2',
            expandFlag: true,
            callbackFlag:false});
        treeTable.expandNode('blade_testCase_tree', {index: '0-2-5',
            expandFlag: true,
            callbackFlag:false});
    })

已知：
界面首次渲染时会展开父节点，所以前两层展开'0','0-2'展开是正常的 ，但要展开第三层的时候失败了

**Comment by LSL1618** [comment]
展开节点的前提是当前节点处于可见状态，也就是说其父节点处于展开状态。展开方法treeTable.expandNode()是异步的，你同时展开3个有上下级关系的节点，无法保证其中的节点在展开的时候一定是可见状态，如果不可见肯定会报错。

**Reply by chenh2h** [reply]
改写代码后能实现按照指定顺序展开树表了
但我还是有些疑问
异步转同步 用async await后 应该就会按照'0','0-2','0-2-5'的顺序展开节点，展开上一个节点后后续节点应该是可见的
但实际上每一次展开需要加100ms的延迟，才能正确展开，是我的异步转同步没有生效吗？求大佬解惑

改写后代码如下

    function asyncExpandNode(time,load_index){
        return new Promise(
            resolve => {
                setTimeout(() => {
                    resolve(treeTable.expandNode('blade_testCase_tree', {index: load_index, expandFlag: true,callbackFlag:false}))
                },time)
            }
        )
    }

    async function syncLoadPath(){
        try{
            await asyncExpandNode(100,'0')
            await asyncExpandNode(100,'0-2')
            await asyncExpandNode(100,'0-2-5')
        }catch(error){
            console.log(error)
        }
    }

    //按钮事件-读取上次路径
    $('#load_last_path').click(function(){
        syncLoadPath()
    })

**Reply by LSL1618** [reply]
你这异步转同步好像不太对劲，你应该在treeTable组件的展开成功之后的回调方法中去执行下一次展开，这样才能保证前后顺序。
